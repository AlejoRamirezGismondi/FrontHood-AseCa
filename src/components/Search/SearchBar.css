

.component-search-input {
    margin: 10px;
}

.component-search-input input {
    border-radius: 4px;
    border: 3px solid #bbb;
    box-sizing: border-box;
    font-size: 18px;
    padding: 10px 8px;
    width: 100%;
    outline: none;
}

.component-search-input input:focus {
    border-color: #bbbbbb;
    box-shadow: 0 0 8px #bbbbbb;
}

