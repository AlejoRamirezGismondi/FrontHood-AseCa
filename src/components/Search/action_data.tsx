import React from 'react';

const actions_mock = {
    "best matches":[
    {
        id: 1,
        action_name: "Adobe",
        symbol: "ADBE",
        company_name: "Adobe Inc",
        price: 488
    },
    {
        id: 2,
        action_name: "Spotify Technology",
        symbol: "SPOT",
        company_name: "Spotify Technology S.A.",
        price: 229
    },
    {
        id: 3,
        action_name: "BJ's Wholesale Club",
        symbol: "BJ",
        company_name: "BJ`s Wholesale Club Holdings Inc",
        price: 45
    },
    {
        id: 4,
        action_name: "Gamestop",
        symbol: "GME",
        company_name: "Gamestop Corporation",
        price: 176
    },
    {
        id: 5,
        action_name: "Tesla",
        symbol: "TSLA",
        company_name: "Tesla Inc",
        price: 580
    },
]}
const actions_mock_search_api = [
    {
    "1. symbol": "ADBE",
    "2. name": "Adobe Inc",
    "9. matchScore": 0.864,
    },
    {
        "1. symbol": "SPOT",
        "2. name": "Spotify Technology S.A.",
        "9. matchScore": 0.875,
    },
    {
        "1. symbol": "BJ",
        "2. name": "BJ`s Wholesale Club Holdings Inc",
        "9. matchScore": 0.756,
    },
    {
        "1. symbol": "GME",
        "2. name": "Gamestop Corporation",
        "9. matchScore": 0.985,
    },
    {
        "1. symbol": "TSLA",
        "2. name": "Tesla Inc",
        "9. matchScore": 0.735,
    },]

export default {actions_mock, actions_mock_search_api}

