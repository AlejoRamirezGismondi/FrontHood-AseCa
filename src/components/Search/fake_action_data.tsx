import React from 'react';



const actions_mock_search_api = {
    "bestMatches": [
        {
            "1. symbol": "TESO",
            "2. name": "Tesco Corporation USA",
            "3. type": "Equity",
            "4. region": "United States",
            "5. marketOpen": "09:30",
            "6. marketClose": "16:00",
            "7. timezone": "UTC-04",
            "8. currency": "USD",
            "9. matchScore": "0.8889"
        },
        {
            "1. symbol": "TSCO.LON",
            "2. name": "Tesco PLC",
            "3. type": "Equity",
            "4. region": "United Kingdom",
            "5. marketOpen": "08:00",
            "6. marketClose": "16:30",
            "7. timezone": "UTC+01",
            "8. currency": "GBX",
            "9. matchScore": "0.7273"
        },
        {
            "1. symbol": "TSCDF",
            "2. name": "Tesco plc",
            "3. type": "Equity",
            "4. region": "United States",
            "5. marketOpen": "09:30",
            "6. marketClose": "16:00",
            "7. timezone": "UTC-04",
            "8. currency": "USD",
            "9. matchScore": "0.7143"
        },
        {
            "1. symbol": "TSCDY",
            "2. name": "Tesco plc",
            "3. type": "Equity",
            "4. region": "United States",
            "5. marketOpen": "09:30",
            "6. marketClose": "16:00",
            "7. timezone": "UTC-04",
            "8. currency": "USD",
            "9. matchScore": "0.7143"
        },
        {
            "1. symbol": "TCO.DEX",
            "2. name": "Tesco PLC",
            "3. type": "Equity",
            "4. region": "XETRA",
            "5. marketOpen": "08:00",
            "6. marketClose": "20:00",
            "7. timezone": "UTC+02",
            "8. currency": "EUR",
            "9. matchScore": "0.7143"
        },
        {
            "1. symbol": "TCO.FRK",
            "2. name": "Tesco PLC",
            "3. type": "Equity",
            "4. region": "Frankfurt",
            "5. marketOpen": "08:00",
            "6. marketClose": "20:00",
            "7. timezone": "UTC+02",
            "8. currency": "EUR",
            "9. matchScore": "0.7143"
        }
    ]
}

const actions_mock = [
    {
    "1. symbol": "ADBE",
    "2. name": "Adobe Inc",
    "9. matchScore": 0.864,
    },
    {
        "1. symbol": "SPOT",
        "2. name": "Spotify Technology S.A.",
        "9. matchScore": 0.875,
    },
    {
        "1. symbol": "BJ",
        "2. name": "BJ`s Wholesale Club Holdings Inc",
        "9. matchScore": 0.756,
    },
    {
        "1. symbol": "GME",
        "2. name": "Gamestop Corporation",
        "9. matchScore": 0.985,
    },
    {
        "1. symbol": "TSLA",
        "2. name": "Tesla Inc",
        "9. matchScore": 0.735,
    },]

export default {actions_mock, actions_mock_search_api}
