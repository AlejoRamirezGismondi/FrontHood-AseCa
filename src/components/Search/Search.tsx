import React, {Component} from "react";
import "./Search.css";
import mock_actions from "./fake_action_data"
import {Link} from "react-router-dom"
import SearchBar from "./SearchBar";
import {get} from "../http"

class Search extends Component {


    constructor(props) {
        super(props);

        this.state = {
            showMessage: false,
            searchInput: "",
            actions: mock_actions.actions_mock
        };
    }
    actionList;
    searchUrl = "search/";

    state = {
        showMessage: false,
        searchInput: "",
        actions: mock_actions.actions_mock
    };

    viewDetails(id) {
        console.log(id)
    }

    handleSearch(){
        console.log(this.state.searchInput)
        if(this.state.searchInput == ""){
            this.setState({showMessage: true})
        } else {
            get(this.searchUrl + this.state.searchInput)
                .then(res => {
                    this.updateActions(res)
                })
        }

    }

    filterBySearchInput(input){
        if(input.length > 1 && input.length <= 30){
            this.setState({ searchInput: input });
        } else {
            this.setState({ searchInput: "" });
        }
    }


    showActions(){
        return this.actionList = this.state.actions.map((item) =>
            <Link to={`/action_detail/${item["1. symbol"]}`}
                  style={{ color: 'inherit', textDecoration: 'none' }}
                  key={item["1. symbol"]}>
                <div className="action-card" data-testid={"action-card-id-"+item["1. symbol"]} key={item["1. symbol"]}
                     onClick={() => this.viewDetails(item["1. symbol"])}>
                    <h2 className="action-card-title">{item["1. symbol"]}</h2>
                    <p className="action-card-company">{item["2. name"]}</p>
                    <p className="action-card-price">Score {item["9. matchScore"]}</p>
                </div>
            </Link>
        )
    }

    updateActions(data){
        this.setState({ actions: data.bestMatches })
        if(this.state.actions.length > 0){
            if(this.state.showMessage === true){
                this.setState({showMessage: false})
            }
        } else {
            if(this.state.showMessage === false){
                this.setState({showMessage: true})
            }
        }
    }

    render() {
        return (
            <div>
                <SearchBar handleSearch={input => this.filterBySearchInput(input)}
                           handleSubmit={() => this.handleSearch()}/>
                <div className="search-unknown"
                     style={{visibility: this.state.showMessage ? 'visible' : 'hidden' }} >
                    Unable to find the action you were looking for
                </div>
                <div className="container" data-testid={"search-container-id"}>
                    {this.showActions()}
                </div>
            </div>
        )
    }
}

export default Search
