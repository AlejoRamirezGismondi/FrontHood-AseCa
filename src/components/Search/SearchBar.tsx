import "./SearchBar.css";
import React from "react";
import SearchIcon from "@material-ui/icons/Search";


function SearchBar(props) {
    return (
        <div className="component-search-input">
            <div className="input-container">
                <input
                    data-testid={"search-input-id"}
                    type="text"
                    id="action-search"
                    placeholder="Search Action"
                    maxLength= {30}
                    minLength= {2}
                    onChange={event => props.handleSearch(event.target.value)}
                />
            </div>
            <button
                type="submit"
                onClick={() => props.handleSubmit()}
                value="Search">
            </button>
        </div>
    )
}

export default SearchBar


