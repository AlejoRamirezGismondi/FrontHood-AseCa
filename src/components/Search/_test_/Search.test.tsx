import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { fireEvent, waitFor } from '@testing-library/react';
import Search from "../Search";
import {BrowserRouter} from 'react-router-dom';
import {getByTestId} from "@testing-library/react";
import exp from "constants";

let root;

beforeEach(() => {
    const div = document.createElement("div");
    ReactDOM.render(
        <BrowserRouter><Search/></BrowserRouter>, div);
    root = div;
})

describe('SearchBar setup', () => {
    it('should render input with place holder Search Action', function () {
        expect(getByTestId(root, "search-input-id")).toHaveProperty("placeholder", "Search Action")
    });

    it('should render input with empty value', function () {
        expect(getByTestId(root, "search-input-id")).toHaveProperty("value", "")
    });

    it('should render input with max length 30 and min length 1', function () {
        const input = getByTestId(root, "search-input-id")
        expect(input).toHaveProperty("maxLength", 30)
        expect(input).toHaveProperty("minLength", 1)
    });
})

describe('Search setup', () => {
    it('should match snapshot', function () {
        //run npm test
        const tree = renderer.create(<BrowserRouter><Search/></BrowserRouter>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('there should be actions in the menu after page finish loading', async () => {
        const container = getByTestId(root, "search-container-id")
        await waitFor(() => expect(container.childElementCount)
            .toBeGreaterThanOrEqual(1))
            .then(() => {
                const action_card = getByTestId(root,"action-card-id-A")
                expect.any(action_card)
            })
    });
})

describe('search', () =>{

    it('should change the value of the input', () => {
        const input = getByTestId(root,"search-input-id")
        // @ts-ignore
        expect(input.value).toBe("")
        fireEvent.change(input, { target: { value: 'ADBE' } })
        // @ts-ignore
        expect(input.value).toBe('ADBE')
    })

    it('should console log value when search button is clicked', function () {
        const input = getByTestId(root,"search-input-id")
        fireEvent.change(input, { target: { value: 'ADBE' } })
        const consoleSpy = jest.spyOn(console, 'log');
        const button = getByTestId(root,"search-button-id")
        fireEvent.click(button);
        expect(consoleSpy).toHaveBeenCalledWith("ADBE");
    });

    it('there should be actions related to input "AB"', async function () {
        const input = getByTestId(root,"search-input-id")
        fireEvent.change(input, { target: { value: 'AB' } })
        const button = getByTestId(root,"search-button-id")
        fireEvent.click(button);
        const container = getByTestId(root, "search-container-id")
        await waitFor(() => expect(container.childElementCount)
            .toBe(2),)
            .then( () => {
                const action_card = getByTestId(root,"action-card-id-A")
                expect.any(action_card)
            })
    });

    it('should show that no actions where found when input "lalala"', async () => {
        const input = getByTestId(root,"search-input-id")
        fireEvent.change(input, { target: { value: 'lalala' } })
        const button = getByTestId(root,"search-button-id")
        fireEvent.click(button);
        const searchUnknown = getByTestId(root, "search-unknown-id")
        expect(searchUnknown.textContent).toBe("")
        await waitFor(() => expect(searchUnknown.textContent)
            .toBe("Unable to find the action you were looking for"),)
    })

})

describe('action cards in Search behaviour', () =>{
    it('should console log something action card is clicked', async function () {
        const consoleSpy = jest.spyOn(console, 'log');
        const container = getByTestId(root, "search-container-id")
        await waitFor(() => expect(container.childElementCount)
            .toBeGreaterThanOrEqual(1),)
            .then( () => {
            const action_card = getByTestId(root,"action-card-id-A")
            fireEvent.click(action_card);
            expect(consoleSpy).toHaveBeenCalledWith("A");
        })
    });

    it('action card should have Symbol and Name', async () => {
        const container = getByTestId(root, "search-container-id")
        await waitFor(() => expect(container.childElementCount)
                .toBeGreaterThanOrEqual(1), )
            .then( () => {
            const actionCard = getByTestId(root,"action-card-id-A");
            const symbol = actionCard.children.item(0).textContent;
            const name = actionCard.children.item(1).textContent;

            expect(symbol).toContain('A');
            expect(name).toContain('Agilent Technologies Inc');
        })
    });

    it('should redirect to specific Action when action-card is clicked', async () => {
        const container = getByTestId(root, "search-container-id")
        const redirectUrl = '/action_detail/Agilent%20Technologies%20Inc/A'

        await waitFor(() => expect(container.childElementCount)
            .toBeGreaterThanOrEqual(1), )
            .then( () => {
                const action_card = getByTestId(root, "action-card-id-A")
                fireEvent.click(action_card);
                expect(window.location.href).toBe(`http://localhost${redirectUrl}`)
            })
    })
})







