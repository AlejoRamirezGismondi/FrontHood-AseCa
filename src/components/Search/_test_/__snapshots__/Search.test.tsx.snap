// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search setup should match snapshot 1`] = `
<div>
  <div
    className="component-search-input"
  >
    <div
      className="input-container"
    >
      <input
        data-testid="search-input-id"
        id="action-search"
        maxLength={30}
        minLength={2}
        onChange={[Function]}
        placeholder="Search Action"
        type="text"
      />
    </div>
    <button
      data-testid="search-button-id"
      onClick={[Function]}
      type="submit"
      value="Search"
    />
  </div>
  <div
    className="search-unknown"
    style={
      Object {
        "visibility": "hidden",
      }
    }
  >
    Unable to find the action you were looking for
  </div>
  <div
    className="container"
    data-testid="search-container-id"
  >
    <a
      href="/action_detail/ADBE"
      onClick={[Function]}
      style={
        Object {
          "color": "inherit",
          "textDecoration": "none",
        }
      }
    >
      <div
        className="action-card"
        data-testid="action-card-id-ADBE"
        onClick={[Function]}
      >
        <h2
          className="action-card-title"
        >
          ADBE
        </h2>
        <p
          className="action-card-company"
        >
          Adobe Inc
        </p>
        <p
          className="action-card-price"
        >
          Score 
          0.864
        </p>
      </div>
    </a>
    <a
      href="/action_detail/SPOT"
      onClick={[Function]}
      style={
        Object {
          "color": "inherit",
          "textDecoration": "none",
        }
      }
    >
      <div
        className="action-card"
        data-testid="action-card-id-SPOT"
        onClick={[Function]}
      >
        <h2
          className="action-card-title"
        >
          SPOT
        </h2>
        <p
          className="action-card-company"
        >
          Spotify Technology S.A.
        </p>
        <p
          className="action-card-price"
        >
          Score 
          0.875
        </p>
      </div>
    </a>
    <a
      href="/action_detail/BJ"
      onClick={[Function]}
      style={
        Object {
          "color": "inherit",
          "textDecoration": "none",
        }
      }
    >
      <div
        className="action-card"
        data-testid="action-card-id-BJ"
        onClick={[Function]}
      >
        <h2
          className="action-card-title"
        >
          BJ
        </h2>
        <p
          className="action-card-company"
        >
          BJ\`s Wholesale Club Holdings Inc
        </p>
        <p
          className="action-card-price"
        >
          Score 
          0.756
        </p>
      </div>
    </a>
    <a
      href="/action_detail/GME"
      onClick={[Function]}
      style={
        Object {
          "color": "inherit",
          "textDecoration": "none",
        }
      }
    >
      <div
        className="action-card"
        data-testid="action-card-id-GME"
        onClick={[Function]}
      >
        <h2
          className="action-card-title"
        >
          GME
        </h2>
        <p
          className="action-card-company"
        >
          Gamestop Corporation
        </p>
        <p
          className="action-card-price"
        >
          Score 
          0.985
        </p>
      </div>
    </a>
    <a
      href="/action_detail/TSLA"
      onClick={[Function]}
      style={
        Object {
          "color": "inherit",
          "textDecoration": "none",
        }
      }
    >
      <div
        className="action-card"
        data-testid="action-card-id-TSLA"
        onClick={[Function]}
      >
        <h2
          className="action-card-title"
        >
          TSLA
        </h2>
        <p
          className="action-card-company"
        >
          Tesla Inc
        </p>
        <p
          className="action-card-price"
        >
          Score 
          0.735
        </p>
      </div>
    </a>
  </div>
</div>
`;
